### PR TITLE
Set default values to map task template

### PR DESCRIPTION
**Pull request by eapolinario**
# TL;DR
Set sane defaults in the map task template, matching the defaults in the [backend](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/array/core/state.go#L154-L160).

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Map tasks containing custom fields would always succeed. That happened because in such cases we were not serializing `min_success_ratio`. In this PR we set defaults to the custom field.

I'm leaving support for `min_successes` intact just in case we have some client out there that's building these array jobs manually using flytekit.

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

**Comment by hamersaw** [comment]
Do we want to set the default concurrency / parallelism to 1? It seems that by default we would want to execute all subtasks concurrently? We could use 0 (or a negative number) to denote that parallelism is not applicable. These would be changes reflected in plugins as well.

This is a similar conversation in setting the default min_success_ratio to set to 1.0 rather than 0.0. I think in this case there is no conceivable scenario where a user would schedule a map task and not allow any tasks to be executed, so 0 is not a valid parallelism.

**Comment by eapolinario** [comment]
> Do we want to set the default concurrency / parallelism to 1? It seems that by default we would want to execute all subtasks concurrently? We could use 0 (or a negative number) to denote that parallelism is not applicable. These would be changes reflected in the flyteplugins as well.

This suggestion looks reasonable. We can leverage the special value `0` to do this and add a check in the client code to ensure that value is never set.

File: flytekit/models/array_job.py [hunk 1]
```diff
@@ -70,13 +70,21 @@ def to_dict(self):
...
         :rtype: dict[T, Text]
         """
-        return _json_format.MessageToDict(
-            _array_job.ArrayJob(
+        array_job = None
+        if self.min_successes is not None:
+            array_job = _array_job.ArrayJob(
                 parallelism=self.parallelism,
                 size=self.size,
                 min_successes=self.min_successes,
```
Comment (kumare3) [inline flytekit/models/array_job.py:78]: is it one-of min_successes and min_success_ratio?

File: flytekit/core/map_task.py [hunk 2]
```diff
@@ -209,7 +209,7 @@ def _raw_execute(self, **kwargs) -> Any:
         return outputs
 
 
-def map_task(task_function: PythonFunctionTask, concurrency: int = None, min_success_ratio: float = None, **kwargs):
+def map_task(task_function: PythonFunctionTask, concurrency: int = 1, min_success_ratio: float = 1.0, **kwargs):
```
Comment (kumare3) [inline flytekit/core/map_task.py:212]: I do not think concurrency should be `1` it should be `0`, which means unbounded?

**Review by kumare3 — COMMENTED** [review]
i think the concurrency should be 0 -> unbounded

File: flytekit/models/array_job.py (hunk 1 above)
Comment (eapolinario) [reply flytekit/models/array_job.py:78]: yes:  https://github.com/flyteorg/flyteidl/blob/091505aae78d935ae55e4d0e46deee5a1e79a22f/protos/flyteidl/plugins/array_job.proto#L20-L29

File: flytekit/core/map_task.py (hunk 2 above)
Comment (eapolinario) [reply flytekit/core/map_task.py:212]: sure thing.

File: flytekit/models/array_job.py [hunk 3]
```diff
@@ -70,13 +70,21 @@ def to_dict(self):
...
             )
-        )
+        elif self.min_success_ratio is not None:
```
Comment (hamersaw) [inline flytekit/models/array_job.py:80]: quick thought, should this be changed to check for 1.0 instead of None? If we default to 1.0 in both flytekit and flyteplugins then there is no reason to write the custom arrayjob if it is 1.0.

Comment (eapolinario) [reply flytekit/models/array_job.py:80]: That is a great alternative IMO.

File: flytekit/models/array_job.py [hunk 4]
```diff
@@ -70,13 +70,24 @@ def to_dict(self):
...
+            array_job = _array_job.ArrayJob(
+                parallelism=self.parallelism,
+                size=self.size,
+                min_success_ratio=self.min_success_ratio,
+            )
+
+        array_job_dict = _json_format.MessageToDict(array_job)
+        if self.parallelism is not None and self.parallelism == 0:
+            array_job_dict['parallelism'] = '0'
```
Comment (eapolinario) [inline flytekit/models/array_job.py:89]: I dislike the fact that we have to write this code to handle the case of `0`. How horrible would it be to use `-1` to represent unbounded instead?

File: flytekit/models/array_job.py (hunk 3 above)
Comment (hamersaw) [reply flytekit/models/array_job.py:80]: I suspect I don't have enough context for where this function is used. It seems to check if the provided values are different then the default to populate the array_job only if necessary. If they are the default values, then we don't write the custom array job parameters and the backend used the default.

I think the 0 actually doesn't mean 0 only applies for parallelism. Though probably unlikely, having a 0.0 min_success_ratio means the map task should succeed even if all subtasks fail, which could be valid. However, a parallelism of 0, meaning no tasks can execute, doesn't make any sense. Do we need to add another check here if parallelism != 0?

Comment (eapolinario) [reply flytekit/models/array_job.py:80]: > I suspect I don't have enough context for where this function is used. It seems to check if the provided values are different then the default to populate the array_job only if necessary. If they are the default values, then we don't write the custom array job parameters and the backend used the default.

This function translates the python values to protobuf, which ends up hitting the issue of [default values](https://developers.google.com/protocol-buffers/docs/proto3#default), specifically how the default value for numeric values is 0. There's no way to set the default value to a different value in a numeric field. 

> 
> I think the 0 actually doesn't mean 0 only applies for parallelism. Though probably unlikely, having a 0.0 min_success_ratio means the map task should succeed even if all subtasks fail, which could be valid. However, a parallelism of 0, meaning no tasks can execute, doesn't make any sense. 

The key here is exactly what you said. In other words, a parallelism of 0 makes sense as a magic value (and not as a user-input). However a min_success_ratio of 0 might make sense as a user-input and since we use a numeric field to represent this in protobuf we cannot distinguish between the two cases in the case of min_success_ratio.

> Do we need to add another check here if parallelism != 0?
IMO, to simplify usage we should allow users to set parallelism to 0 with the caveat tha this means unbounded concurrency. The scenario I'm thinking is someone generating these configurations programmatically, if we we don't allow for parallelism to be set to 0 they will have to special case that in their configuration (which is very annoying).

**Review by hamersaw — APPROVED** [review]
